From f421c3d0f3954392c63c5c4e68712eaf9d493167 Mon Sep 17 00:00:00 2001
From: Alastair Hughes <hobbitalastair@gmail.com>
Date: Fri, 19 Aug 2016 19:20:45 +1200
Subject: [PATCH] Revert "Always create dynamic sections for -E/--dynamic-list"

This reverts commit bf89386a862ace008f0152bca8bddf996d3993c8, which
breaks the static gcc build when compiling the host gcc as -rdynamic
gets passed to gcc, which in turn passes -export-dynamic to the linker.
This means that cc1 and cc1plus are built as dynamic executables,
despite there being no shared libraries or a linker/loader available.
---
 bfd/elflink.c                  |  7 ++-----
 ld/testsuite/ld-elf/pr19617.s  | 19 -------------------
 ld/testsuite/ld-elf/pr19617a.d | 13 -------------
 ld/testsuite/ld-elf/pr19617b.d | 11 -----------
 ld/testsuite/ld-elf/pr19617c.d |  9 ---------
 5 files changed, 2 insertions(+), 57 deletions(-)
 delete mode 100644 ld/testsuite/ld-elf/pr19617.s
 delete mode 100644 ld/testsuite/ld-elf/pr19617a.d
 delete mode 100644 ld/testsuite/ld-elf/pr19617b.d
 delete mode 100644 ld/testsuite/ld-elf/pr19617c.d

diff --git a/bfd/elflink.c b/bfd/elflink.c
index 9e9a33c..8ac209d 100644
--- a/bfd/elflink.c
+++ b/bfd/elflink.c
@@ -3745,14 +3745,11 @@ elf_link_add_object_symbols (bfd *abfd, struct bfd_link_info *info)
       /* If we are creating a shared library, create all the dynamic
 	 sections immediately.  We need to attach them to something,
 	 so we attach them to this BFD, provided it is the right
-	 format and is not from ld --just-symbols.  Always create the
-	 dynamic sections for -E/--dynamic-list.  FIXME: If there
+	 format and is not from ld --just-symbols.  FIXME: If there
 	 are no input BFD's of the same format as the output, we can't
 	 make a shared library.  */
       if (!just_syms
-	  && (bfd_link_pic (info)
-	      || (!bfd_link_relocatable (info)
-		  && (info->export_dynamic || info->dynamic)))
+	  && bfd_link_pic (info)
 	  && is_elf_hash_table (htab)
 	  && info->output_bfd->xvec == abfd->xvec
 	  && !htab->dynamic_sections_created)
diff --git a/ld/testsuite/ld-elf/pr19617.s b/ld/testsuite/ld-elf/pr19617.s
deleted file mode 100644
index 0acb8f5..0000000
--- a/ld/testsuite/ld-elf/pr19617.s
+++ /dev/null
@@ -1,19 +0,0 @@
-	.data
-	.global bar
-bar:
-	.type bar,"object"
-	.long 0
-	.text
-	.type start,"function"
-	.global start
-start:
-	.type _start,"function"
-	.global _start
-_start:
-	.type __start,"function"
-	.global __start
-__start:
-	.type main,"function"
-	.global main
-main:
-	.long 0
diff --git a/ld/testsuite/ld-elf/pr19617a.d b/ld/testsuite/ld-elf/pr19617a.d
deleted file mode 100644
index 784aacb..0000000
--- a/ld/testsuite/ld-elf/pr19617a.d
+++ /dev/null
@@ -1,13 +0,0 @@
-#source: pr19617.s
-#ld: -E
-#readelf : --dyn-syms --wide
-#target: *-*-linux* *-*-gnu* *-*-solaris*
-
-Symbol table '\.dynsym' contains [0-9]+ entries:
- +Num: +Value +Size Type +Bind +Vis +Ndx Name
- +0: 0+ +0 +NOTYPE +LOCAL +DEFAULT +UND +
-#...
- +[0-9]+: +[a-f0-9]+ +0 +FUNC +GLOBAL +DEFAULT +[0-9]+ +start
-#...
- +[0-9]+: +[a-f0-9]+ +0 +OBJECT +GLOBAL +DEFAULT +[0-9]+ +bar
-#pass
diff --git a/ld/testsuite/ld-elf/pr19617b.d b/ld/testsuite/ld-elf/pr19617b.d
deleted file mode 100644
index e1dcb71..0000000
--- a/ld/testsuite/ld-elf/pr19617b.d
+++ /dev/null
@@ -1,11 +0,0 @@
-#source: pr19617.s
-#ld: --dynamic-list-data
-#readelf : --dyn-syms --wide
-#target: *-*-linux* *-*-gnu* *-*-solaris*
-
-Symbol table '\.dynsym' contains [0-9]+ entries:
- +Num: +Value +Size Type +Bind +Vis +Ndx Name
- +0: 0+ +0 +NOTYPE +LOCAL +DEFAULT +UND +
-#...
- +[0-9]+: +[a-f0-9]+ +0 +OBJECT +GLOBAL +DEFAULT +[0-9]+ +bar
-#pass
diff --git a/ld/testsuite/ld-elf/pr19617c.d b/ld/testsuite/ld-elf/pr19617c.d
deleted file mode 100644
index 00e2e7e..0000000
--- a/ld/testsuite/ld-elf/pr19617c.d
+++ /dev/null
@@ -1,9 +0,0 @@
-#source: pr19617.s
-#ld: --dynamic-list-data
-#readelf : --dyn-syms --wide
-#target: *-*-linux* *-*-gnu* *-*-solaris*
-
-#failif
-#...
- +[0-9]+: +[a-f0-9]+ +0 +FUNC +GLOBAL +DEFAULT +[0-9]+ +start
-#...
-- 
2.9.3

